lib/Parse/Lexer.cpp

```diff
@@ -1991,12 +1996,103 @@ bool Lexer::tryLexRegexLiteral(const char *TokStart) {
     return true;
   }
 
-  // Otherwise, we either had a successful lex, or something that was
-  // recoverable.
+  // If we're lexing `/.../`, error if we ended on the opening of a comment.
+  // We prefer to lex the comment as it's more likely than not that is what
+  // the user is expecting.
+  // TODO: This should be sunk into the Swift library.
+  if (*TokStart == '/' && (*CurPtr == '*' || *CurPtr == '/')) {
+    diagnose(TokStart, diag::lex_regex_literal_unterminated);
+
+    // Move the pointer back to the '/' of the comment.
+    CurPtr--;
+  }
+
+  // We either had a successful lex, or something that was recoverable.
   formToken(tok::regex_literal, TokStart);
   return true;
 }
 
+bool Lexer::tryLexAsForwardSlashRegexLiteral(State S, bool mustBeRegex) {
+  if (!LangOpts.EnableForwardSlashRegexLiterals)
+    return false;
+
+  auto priorState = getStateForBeginningOfToken(NextToken, LeadingTrivia);
+
+  // Re-lex from the given state.
+  restoreState(S);
+
+  // While we restored state, that would have re-advanced the lexer. This is
+  // good in that it's filled in all the interesting properties of the token
+  // (trivia, is on new line, etc), but we need to rewind to re-lex the actual
+  // kind.
+  auto *TokStart = getBufferPtrForSourceLoc(NextToken.getLoc());
+  assert(*TokStart == '/');
+  CurPtr = TokStart + 1;
```

Comment:
I think this is not ideal, but it's OK.

But, I would just introduce a flag like `ShouldTryRegexLiteralOnSlash` in `Lexer`, then
in `lexImpl()`
```cpp
case '/':
  if (ShouldTryRegexLiteralOnSlash && tryLexRegexLiteral(TokStart)) {
    return;
  }
  ... comment thing ...
  return lexOperatorIdentifier();
```
And here:
```cpp
llvm::SaveAndRestore<bool> Tmp(ShouldTryRegexLiteralOnSlash, true);
restoreState(S);
// 'restoreState()' calls 'lexImpl()' so 'NextToken' should be 'regex_literal' now (if it can be).
```

As for `mustBeRegex` thing, I think you could just check `*getBufferPtrForSourceLoc(S.Loc)`